dali/kernels/signal/wavelet/mother_wavelet.cu

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cmath>
+#include <vector>
+#include "dali/kernels/signal/wavelet/mother_wavelet.cuh"
+#include "dali/core/math_util.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+template <typename T>
+HaarWavelet<T>::HaarWavelet(const std::vector<T> &args) {
+  if (args.size() != 0) {
+    throw new std::invalid_argument("HaarWavelet doesn't accept any arguments.");
+  }
+}
+
+template <typename T>
+__device__ T HaarWavelet<T>::operator()(const T &t) const {
+  if (0.0 <= t && t < 0.5) {
+    return 1.0;
+  }
+  if (0.5 <= t && t < 1.0) {
+    return -1.0;
+  }
+  return 0.0;
+}
+
+template class HaarWavelet<float>;
+template class HaarWavelet<double>;
+
+template <typename T>
+GaussianWavelet<T>::GaussianWavelet(const std::vector<T> &args) {
+  if (args.size() != 1) {
+    throw new std::invalid_argument("GaussianWavelet accepts exactly one argument - n.");
+  }
+  if (args[0] < 1.0 || args[0] > 8.0) {
+    throw new std::invalid_argument(
+      "GaussianWavelet's argument n should be integer from range [1,8].");
+  }
+  this->n = args[0];
+}
+
+template <typename T>
+__device__ T GaussianWavelet<T>::operator()(const T &t) const {
+  T expTerm = std::exp(-std::pow(t, 2.0));
+  T sqrtTerm = 1.2533141373155001;  // std::sqrt(M_PI/2.0)
+  switch (static_cast<int>(n)) {
+  case 1:
```

Comment:
I think it would be less error prone to use enums from wavelet_name.h here.

Reply:
I'm not sure if I understand. The parameter n here is the order of the derivative of Gaussian wavelet. I have defined functions for 8 possible values since that's what most implementations do. Do you propose to create another enum for this parameter n?

Reply:
My bad. I confused it with with the Wavelet type.